drivers/usb/udc/udc_smartbond.c

```diff
@@ -0,0 +1,1784 @@
+{
+	struct usb_dc_state *dev_state = CONTAINER_OF(item, struct usb_dc_state, ep0_setup_work);
+	struct net_buf *buf;
+	int err;
+	const struct device *dev = dev_state->dev;
+	struct smartbond_ep_state *ep0_out_state;
+
+	buf = udc_ctrl_alloc(dev, USB_CONTROL_EP_OUT, sizeof(struct usb_setup_packet));
+	if (buf == NULL) {
+		LOG_ERR("Failed to allocate for setup");
+		return;
+	}
+
+	udc_ep_buf_set_setup(buf);
+	net_buf_add_mem(buf, dev_state->setup_buffer, 8);
+	ep0_out_state = EP0_OUT_STATE(dev_state);
+	ep0_out_state->last_packet_size = 0;
+	ep0_out_state->buf = NULL;
+	udc_ctrl_update_stage(dev, buf);
+
+	if (udc_ctrl_stage_is_data_out(dev)) {
+		/*  Allocate and feed buffer for data OUT stage */
+		LOG_DBG("s:%p|feed for -out-", buf);
+		err = usbd_ctrl_feed_dout(dev, udc_data_stage_length(buf));
+		if (err == -ENOMEM) {
+			err = udc_submit_ep_event(dev, buf, err);
+		}
+	} else if (udc_ctrl_stage_is_data_in(dev)) {
+		/* Allocate buffer for Status OUT state */
+		err = usb_dc_smartbond_alloc_status_out(dev);
+		if (err == -ENOMEM) {
+			err = udc_submit_ep_event(dev, buf, err);
+		} else {
+			err = udc_ctrl_submit_s_in_status(dev);
+			if (err == -ENOMEM) {
+				err = udc_submit_ep_event(dev, buf, err);
+			}
+		}
+	} else {
+		err = udc_ctrl_submit_s_status(dev);
+	}
+}
+
+static int udc_smartbond_enable(const struct device *dev)
+{
+	struct usb_dc_state *dev_state = dev->data;
+	const struct udc_smartbond_config *cfg = dev->config;
+
+	LOG_DBG("Enable UDC");
+
+	usb_change_state(dev_state, true, dev_state->vbus_present);
+
+	if (udc_ep_enable_internal(dev, USB_CONTROL_EP_OUT,
+				   USB_EP_TYPE_CONTROL, 8, 0)) {
+		LOG_ERR("Failed to enable control endpoint");
+		return -EIO;
+	}
+
+	if (udc_ep_enable_internal(dev, USB_CONTROL_EP_IN,
+				   USB_EP_TYPE_CONTROL, 8, 0)) {
+		LOG_ERR("Failed to enable control endpoint");
+		return -EIO;
+	}
+
+	irq_enable(cfg->udc_irq);
+
+	return 0;
+}
+
+static int udc_smartbond_disable(const struct device *dev)
+{
+	struct usb_dc_state *dev_state = dev->data;
+	const struct udc_smartbond_config *cfg = dev->config;
+
+	LOG_DBG("Disable UDC");
+
+	usb_change_state(dev_state, false, dev_state->vbus_present);
+
+	irq_disable(cfg->udc_irq);
+
+	return 0;
+}
+
+/*
+ * Prepare and configure most of the parts, if the controller has a way
+ * of detecting VBUS activity it should be enabled here.
+ * Only udc_smartbond_enable() makes device visible to the host.
+ */
+static int udc_smartbond_init(const struct device *dev)
+{
+	struct udc_data *data = dev->data;
+	struct usb_dc_state *dev_state = dev->data;
+	const struct udc_smartbond_config *cfg = dev->config;
+	struct smartbond_ep_reg_set *reg_set = (struct smartbond_ep_reg_set *)&(USB->USB_EPC0_REG);
+	const uint16_t mps = 1023;
+	int err;
+
+	dev_state->dev = dev;
+
+	/*
+	 * You do not need to initialize it if your driver does not use
+	 * udc_lock_internal() / udc_unlock_internal(), but implements its
+	 * own mechanism.
+	 */
```

Comment:
Looks like stale template comment that should be removed.

Reply:
comment removed